test/e2e/httpproxy/ip_filtering_test.go

```diff
@@ -0,0 +1,399 @@
+
+		// Deny all ips so that the next request fails
+		require.NoError(t, retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			if err := f.Client.Get(ctx, client.ObjectKeyFromObject(p), p); err != nil {
+				return err
+			}
+
+			p.Spec.VirtualHost.IPDenyFilterPolicy = []contourv1.IPFilterPolicy{
+				{
+					Source: contourv1.IPFilterSourcePeer,
+					CIDR:   "10.8.8.8/0",
+				},
+				{
+					Source: contourv1.IPFilterSourceRemote,
+					CIDR:   "10.8.8.8/0",
+				},
+			}
+
+			return f.Client.Update(ctx, p)
+		}))
+
+		// Make a request against the proxy, it should fail
+		res, ok = f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+			Path:      "/one",
+			Host:      p.Spec.VirtualHost.Fqdn,
+			Condition: e2e.HasStatusCode(403),
+		})
+		require.NotNil(t, res, "request never succeeded")
+		require.Truef(t, ok, "expected 403 response code, got %d", res.StatusCode)
+
+		// Allow requests from 10.10.10.10 on the route
+		require.NoError(t, retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+			if err := f.Client.Get(ctx, client.ObjectKeyFromObject(p), p); err != nil {
+				return err
+			}
+
+			p.Spec.Routes[0].IPAllowFilterPolicy = []contourv1.IPFilterPolicy{
+				{
+					Source: contourv1.IPFilterSourceRemote,
+					CIDR:   "10.10.10.10",
+				},
+			}
+			p.Spec.Routes[0].IPDenyFilterPolicy = nil
+
+			return f.Client.Update(ctx, p)
+		}))
+
+		// Add an X-Forwarded-For header to match the allowed ip, it should succeed
+		res, ok = f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+			Path:      "/one",
+			Host:      p.Spec.VirtualHost.Fqdn,
+			Condition: e2e.HasStatusCode(200),
+			RequestOpts: []func(*http.Request){
+				e2e.OptSetHeaders(map[string]string{"X-Forwarded-For": "10.10.10.10"}),
+			},
+		})
+		require.NotNil(t, res, "request never succeeded")
+		require.Truef(t, ok, "expected 200 response code, got %d", res.StatusCode)
+
+		// A request with the wrong ip should fail
+		res, ok = f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+			Path:      "/one",
+			Host:      p.Spec.VirtualHost.Fqdn,
+			Condition: e2e.HasStatusCode(403),
+			RequestOpts: []func(*http.Request){
+				e2e.OptSetHeaders(map[string]string{"X-Forwarded-For": "10.10.10.0"}),
+			},
+		})
+		require.NotNil(t, res, "request never succeeded")
+		require.Truef(t, ok, "expected 403 response code, got %d", res.StatusCode)
+
+		// A request against the other route should fail (virtualhost-level filter applies)
+		res, ok = f.HTTP.RequestUntil(&e2e.HTTPRequestOpts{
+			Path:      "/other",
+			Host:      p.Spec.VirtualHost.Fqdn,
+			Condition: e2e.HasStatusCode(403),
+			RequestOpts: []func(*http.Request){
+				e2e.OptSetHeaders(map[string]string{"X-Forwarded-For": "10.10.10.0"}),
+			},
+		})
+		require.NotNil(t, res, "request never succeeded")
+		require.Truef(t, ok, "expected 403 response code, got %d", res.StatusCode)
+	})
+
+	Specify("requests can be filtered by ip address in included routes", func() {
```

Comment:
maybe better as a featuretest, but can we add a test where both the root and child proxy have ip allow/deny rules? who takes precedence in that case/what are the semantics there?

Reply:
> but can we add a test where both the root and child proxy have ip allow/deny rules? who takes precedence in that case/what are the semantics there?

Only the parent httproxy can have virtualhost-level rules, and any rules in child httpproxies will be per-route (and unique, otherwise validation on the overall httproxy would fail).

There's no contour-level de-tangling of ip filter rules if that makes sense

Reply:
From my testing, it looks like the presence of any ip allow/deny policies at the route level completely override any vhost-level settings, so if you specify anything for a route, the vhost-level settings are effectively ignored for that route. There's no merging.

Reply:
maybe this result need into projectcoutour document.

Reply:
There's a section in the user docs for this now: https://github.com/projectcontour/contour/blob/eff7460f1389a2e1b63a82eda33d4379f91188e4/site/content/docs/main/config/ip-filtering.md